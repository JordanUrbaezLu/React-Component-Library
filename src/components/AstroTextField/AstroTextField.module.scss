@use "../../utility/globals";

.astroTextFieldContainer {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
}

.astroTextField {
  outline: none !important;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
}

.primary {
  border: 3px solid globals.$primary-color;
  &:focus {
    border: 3px solid globals.$primary-color-pseudo-class;
  }
}

.secondary {
  border: 3px solid globals.$secondary-color;
  &:focus {
    border: 3px solid globals.$secondary-color-pseudo-class;
  }
}

.tertiary {
  border: 3px solid globals.$tertiary-color;
  &:focus {
    border: 3px solid globals.$tertiary-color-pseudo-class;
  }
}

.small {
  border-radius: 6px;
  font-size: 12px;
  padding: 4px;
}

.medium {
  border-radius: 8px;
  font-size: 18px;
  padding: 6px;
}

.large {
  border-radius: 10px;
  font-size: 24px;
  padding: 8px;
}

.iconSmall {
  font-size: 21px;
  padding: 1px;
  display: flex;
  align-items: center;
  cursor: pointer;
}

.iconMedium {
  font-size: 28px;
  padding: 3px;
  display: flex;
  align-items: center;
  cursor: pointer;
}

.iconLarge {
  font-size: 35px;
  padding: 5px;
  display: flex;
  align-items: center;
  cursor: pointer;
}

.disabled {
  border: none !important;
  cursor: default !important;
}
