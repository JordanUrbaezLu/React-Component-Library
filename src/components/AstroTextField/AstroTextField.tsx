import classNames from "classnames";
import * as React from "react";
// TODO style needs some work, text over flows into icon, maybe move icon outside of input box
import styles from "./AstroTextField.module.scss";

export type AstroTextFieldVariant = "primary" | "secondary" | "tertiary";
export type AstroTextFieldSize = "small" | "medium" | "large";
export type AstroTextFieldType =
  | "email"
  | "number"
  | "password"
  | "search"
  | "tel"
  | "text"
  | "time"
  | "url";

interface AstroTextFieldProps {
  variant?: AstroTextFieldVariant;
  size?: AstroTextFieldSize;
  placeholder?: string;
  trailingIcon?: React.ReactNode;
  disabled?: boolean;
  type?: AstroTextFieldType;
  label?: string;
}

const AstroTextField: React.FC<AstroTextFieldProps> = ({
  variant = "primary",
  size = "medium",
  placeholder,
  trailingIcon,
  disabled = false,
  type = "text",
  label = "textfield",
}) => {
  const classes = classNames(
    styles.astroTextField,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    variant === "tertiary" && styles.tertiary,
    size === "small" && styles.small,
    size === "medium" && styles.medium,
    size === "large" && styles.large,
    disabled && styles.disabled
  );

  const iconClasses = classNames(
    size === "small" && styles.iconSmall,
    size === "medium" && styles.iconMedium,
    size === "large" && styles.iconLarge,
    disabled && styles.disabled
  );

  return (
    <div className={styles.astroTextFieldContainer}>
      <input
        aria-label={label}
        type={type}
        disabled={disabled}
        className={classes}
        placeholder={placeholder}
      />
      {trailingIcon && <span className={iconClasses}>{trailingIcon}</span>}
    </div>
  );
};

export default AstroTextField;
