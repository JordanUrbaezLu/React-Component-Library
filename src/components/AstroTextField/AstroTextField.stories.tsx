import { ComponentStory, ComponentMeta } from "@storybook/react";
import { AiFillEye, AiOutlineArrowRight } from "react-icons/ai";
import AstroTextField from "./AstroTextField";

export default {
  title: "AstroTextField",
  component: AstroTextField,
} as ComponentMeta<typeof AstroTextField>;

const Template: ComponentStory<typeof AstroTextField> = (args) => {
  return <AstroTextField {...args} />;
};

export const Default = Template.bind({});

export const Primary = Template.bind({});
Primary.args = {
  variant: "primary",
};

export const Secondary = Template.bind({});
Secondary.args = {
  variant: "secondary",
};

export const Tertiary = Template.bind({});
Tertiary.args = {
  variant: "tertiary",
};

export const Small = Template.bind({});
Small.args = {
  size: "small",
};

export const Medium = Template.bind({});
Medium.args = {
  size: "medium",
};

export const Large = Template.bind({});
Large.args = {
  size: "large",
};

export const Placeholder = Template.bind({});
Placeholder.args = {
    placeholder: "Search..."
}

// disabled state needs better design
export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

// some bug where if the text gets too long it interacts with the trailing icon
export const TrailingIcon = Template.bind({});
TrailingIcon.args = {
  trailingIcon: <AiFillEye />,
};

export const PasswordType = Template.bind({});
PasswordType.args = {
  type: "password",
};

// this is broken, maybe because value isn't set, maybe not broken?
export const EmailType = Template.bind({});
EmailType.args = {
  type: "email",
};
