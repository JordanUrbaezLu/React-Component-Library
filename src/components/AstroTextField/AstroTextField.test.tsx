import { render, screen } from "@testing-library/react";
import AstroTextField, {
  AstroTextFieldSize,
  AstroTextFieldVariant,
  AstroTextFieldType,
} from "./AstroTextField";
import { axe, toHaveNoViolations } from "jest-axe";
import "@testing-library/jest-dom";
expect.extend(toHaveNoViolations);

describe("AstroTextField", () => {
  // Fixes some jest-axe issues
  const { getComputedStyle } = window;
  window.getComputedStyle = (elt) => getComputedStyle(elt);
  // Fixes some jest-axe issues
  test("Should render component correctly.", () => {
    const { container } = render(<AstroTextField />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(<AstroTextField />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test.each<AstroTextFieldVariant>(["primary", "secondary", "tertiary"])(
    "Should render variants correctly.",
    (variant) => {
      render(<AstroTextField variant={variant} />);
      expect(screen.getByLabelText("textfield")).toHaveClass(variant);
    }
  );

  test.each<AstroTextFieldSize>(["small", "medium", "large"])(
    "Should render size correctly.",
    (size) => {
      render(<AstroTextField size={size} />);
      expect(screen.getByLabelText("textfield")).toHaveClass(size);
    }
  );

  test("Should render disabled correctly.", () => {
    render(<AstroTextField disabled />);
    expect(screen.getByLabelText("textfield")).toHaveClass("disabled");
  });

  test("Should render label correctly.", () => {
    render(<AstroTextField label="astro" />);
    expect(screen.getByLabelText("astro")).toBeInTheDocument();
  });

  test.each<AstroTextFieldType>([
    "email",
    "number",
    "password",
    "search",
    "tel",
    "text",
    "time",
    "url",
  ])("Should render type correctly.", (type) => {
    render(<AstroTextField type={type} />);
    expect(screen.getByLabelText("textfield")).toHaveAttribute("type", type);
  });

  test("Should render trailing icon correctly.", () => {
    render(<AstroTextField trailingIcon="trailing" />);
    expect(screen.getByText("trailing")).toBeInTheDocument();
  });
});
