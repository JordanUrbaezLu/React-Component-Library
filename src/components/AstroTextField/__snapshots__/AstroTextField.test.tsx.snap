// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstroTextField Should render component correctly. 1`] = `
<div
  class="astroTextFieldContainer"
>
  <input
    aria-label="textfield"
    class="astroTextField primary medium"
    type="text"
  />
</div>
`;
